datafu-spark/src/main/resources/pyspark_utils/df_utils.py

```diff
@@ -0,0 +1,171 @@
+    splits both of the DFs to two parts according to the skewed keys.
+    1. Map-join: broadcasts the skewed-keys part of the not skewed DF to the skewed-keys part of the skewed DF
+    2. Regular join: between the remaining two parts.
+    :param not_skewed_df: not skewed DataFrame
+    :param skewed_df: skewed DataFrame
+    :param join_col: join column
+    :param number_of_custs_to_broadcast: number of custs to broadcast
+    :return: DataFrame representing the data after the operation
+    """
+    jdf = _get_utils(skewed_df).broadcastJoinSkewed(not_skewed_df._jdf, skewed_df._jdf, join_col, number_of_custs_to_broadcast)
+    return DataFrame(jdf, not_skewed_df.sql_ctx)
+
+
+def join_with_range(df_single, col_single, df_range, col_range_start, col_range_end, decrease_factor):
+    """
+    Helper function to join a table with column to a table with range of the same column.
+    For example, ip table with whois data that has range of ips as lines.
+    The main problem which this handles is doing naive explode on the range can result in huge table.
+    requires:
+    1. single table needs to be distinct on the join column, because there could be a few corresponding ranges so we dedup at the end - we choose the minimal range.
+    2. the range and single columns to be numeric.
+    """
+    jdf = _get_utils(df_single).joinWithRange(df_single._jdf, col_single, df_range._jdf, col_range_start, col_range_end, decrease_factor)
+    return DataFrame(jdf, df_single.sql_ctx)
+
+
+def join_with_range_and_dedup(df_single, col_single, df_range, col_range_start, col_range_end, decrease_factor, dedup_small_range):
+    """
+    Helper function to join a table with column to a table with range of the same column.
+    For example, ip table with whois data that has range of ips as lines.
+    The main problem which this handles is doing naive explode on the range can result in huge table.
+    requires:
+    1. single table needs to be distinct on the join column, because there could be a few corresponding ranges so we dedup at the end - we choose the minimal range.
+    2. the range and single columns to be numeric.
+    """
+    jdf = _get_utils(df_single).joinWithRangeAndDedup(df_single._jdf, col_single, df_range._jdf, col_range_start, col_range_end, decrease_factor, dedup_small_range)
+    return DataFrame(jdf, df_single.sql_ctx)
+
+
+def _cols_to_java_cols(cols):
+    return _map_if_needed(lambda x: x._jc, cols)
+
+
+def _dfs_to_java_dfs(dfs):
+    return _map_if_needed(lambda x: x._jdf, dfs)
+
+
+def _map_if_needed(func, itr):
+    return map(func, itr) if itr is not None else itr
+
+
+def activate():
+    """Activate integration between datafu-spark and PySpark.
+    This function only needs to be called once.
+
+    This technique taken from pymongo_spark
+    https://github.com/mongodb/mongo-hadoop/blob/master/spark/src/main/python/pymongo_spark.py
+    """
+    pyspark.sql.DataFrame.dedup = dedup
```

Comment:
I sort of wonder, and I'm not sure, but maybe we should add a package to make it `pyspark.sql.DataFrame.datafu.dedup`? I'm not sure. @Matthayes what do you think? My concern is name collision as they add features, which they are wont to do.

Reply:
as we've changed the names of the short methods i don't think we have a high risk for collisions..

Reply:
Awesome!